pipeline/metadata/ip_metadata_chooser.py

```diff
@@ -0,0 +1,104 @@
+"""A module which contains a number of sources of ip metadata and chooses how to add them to ips."""
+import datetime
+import logging
+from typing import Type
+
+from pipeline.metadata import caida_ip_metadata
+from pipeline.metadata import dbip
+from pipeline.metadata import maxmind
+from pipeline.metadata.flatten import Row
+
+
+# pylint: disable=too-many-instance-attributes
+class IpMetadataChooser():
+  """Business logic for selecting IP metadata values from a number of sources"""
+
+  def __init__(
+      self,
+      date: datetime.date,
+      caida_ip_metadata_class: Type[caida_ip_metadata.CaidaIpMetadata],
+      caida_ip_metadata_bucket_folder: str,
+      # pylint: disable=unused-argument
+      maxmind_class: Type[maxmind.MaxmindIpMetadata],
+      maxmind_bucket_folder: str,
+      # pylint: enable=unused-argument
+      dbip_class: Type[dbip.DbipMetadata],
+      dbip_bucket_folder: str,
+  ) -> None:
+    """Store all metadata sources for future querying
+
+    Args:
+      date: date to initialize caida metadata at
+      caida_ip_metadata_class: an IpMetadataInterface subclass (class, not instance)
+      caida_ip_metadata_bucket_folder: gcs folder with CAIDA ip metadata files
+      maxmind_class: an IpMetadataInterface subclass (class, not instance)
+      maxmind_bucket_folder: gcs folder with maxmind files
+      dbip_class: a DbipMetadata class (class, not instance)
+      dbip_bucket_folder: gcs folder with dbip files
+    """
+    self.caida = caida_ip_metadata_class(date, caida_ip_metadata_bucket_folder,
```

Comment:
Move these to the factory function and pass the caida, maxmind and dbip objects in the constructor, instead of class + bucket. No need for this class to know how they are created.

Reply:
done